daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import org.apache.daffodil.dpath.NodeInfo.PrimType
+
+trait CommonContextView {
+  def namespaces: scala.xml.NamespaceBinding
+}
+
+trait TermView extends CommonContextView {
+  def isArray: Boolean
+  def isOptional: Boolean
+  def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
```

Comment:
Do we want these walkDSOM function to be protected? Is there a reason that user might want to call walkDSOM directrly? We can also make them public in the future if needed, but it's much harder to switch to make them protected if they are already public.

Reply:
The walk is currently initiated here:
`final def walkFromRoot(schemaSetRoot: RootView): T = {
    onWalkBegin(schemaSetRoot)
    schemaSetRoot.walkDSOM(this)
    // onWalkEnd will return the value returned by this method!
    onWalkEnd(schemaSetRoot)
  }`
As you can see, this method requires `walkDSOM` to be a public member of `TermView`.  I am not quite sure how exactly the walk would begin without this method being public.

Reply:
Ah, that's right. I guess the only way to do it would be to have the walkDSOM function not be on the View but be on the actual DSOM SchemaComponents, and all the walking occurs on the internal/private/unstable API. Unclear if that adds much benefit though. Perhaps there's value in allowing the implementation to change how the DSOM is walked.

Reply:
There are different walk patterns yes. So I think keeping the walkers on the traits where they can be overridden or new ones added makes sense.

Reply:
Actually, these walkDSOM functions can't really be easily modified since the objects that implement the *View traits are provided by Daffodil. If we want the user to be able to implement custom walking logic, we probably need to go back something like how it was before where the Walker has the walk logic rather than the Views.

Reply:
I could always change `walkFromRoot` to a non-final method.  This would allow the user to override it and then instead of doing `schemaSetRoot.walkDSOM(this);` they could call their own custom helper method, similar to what we had before.  Keeping the current code intact while doing this would enable the existing walking strategy to be the default while also allowing the user to override with their own implementation.

Reply:
Seems reasonable. So it sounds like basically the API becomes all the different Views and their members, plus some new public method to get the RootView. WIth the RootView, the user can traverse the DSOM tree however they want. And then we have the recursive DSOM walker implementation for users that just want to get a bunch of events and not have to worry about how things DSOM is walked. This feel like it makes the API a little larger and could make it difficult to change in the future, but as long as its experimental until it seems stable that's fine with me.

Reply:
This would make the API larger due to this extra flexibility.  I went ahead and removed the `final` modifier from this method in the latest revision.